### PR TITLE
Use Gradle profiler for running cross build tests

### PR DESCRIPTION
**Pull request by wolfs**
Fixes https://github.com/gradle/gradle/issues/10540

### CODE REVIEW

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractCrossBuildPerformanceTestRunner.groovy [hunk 1]
```diff
@@ -36,7 +36,7 @@ abstract class AbstractCrossBuildPerformanceTestRunner extends AbstractGradleBui
     protected void defaultSpec(BuildExperimentSpec.Builder builder) {
         super.defaultSpec(builder)
         if (builder instanceof GradleBuildExperimentSpec.GradleBuilder) {
-            builder.invocation.distribution(gradleDistribution)
+            ((GradleBuildExperimentSpec.GradleBuilder) builder).invocation.distribution(gradleDistribution)
```
Comment (lptr) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractCrossBuildPerformanceTestRunner.groovy]: Why is this cast needed? Shouldn't Groovy treat the `if` clause as if `builder` was already cast?

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractGradleBuildPerformanceTestRunner.groovy [hunk 2]
```diff
@@ -83,10 +72,6 @@ abstract class AbstractGradleBuildPerformanceTestRunner<R extends PerformanceTes
     }
 
     protected void defaultSpec(BuildExperimentSpec.Builder builder) {
-        builder.setListener(buildExperimentListener)
-        builder.setInvocationCustomizer(invocationCustomizer)
-        builder.setBuildMutators(buildMutators)
-        builder.setMeasuredBuildOperations(ImmutableList.copyOf(measuredBuildOperations))
     }
```
Comment (lptr) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractGradleBuildPerformanceTestRunner.groovy]: Should this be `abstract` now?

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/MavenBuildExperimentSpec.groovy [hunk 3]
```diff
@@ -92,6 +90,26 @@ class MavenBuildExperimentSpec extends BuildExperimentSpec {
...
 
+        @Override
+        void setBuildMutators(ImmutableList<Function<InvocationSettings, BuildMutator>> mutators) {
+            throw new UnsupportedOperationException("Maven for Gradle profiler is not yet supported")
+        }
+
+        @Override
+        ImmutableList<Function<InvocationSettings, BuildMutator>> getBuildMutators() {
+            return null
+        }
```
Comment (lptr) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/MavenBuildExperimentSpec.groovy]: Should this throw `UOE` instead? Or return an empty list perhaps?

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/experiment/buildcache/LocalTaskOutputCacheCrossBuildPerformanceTest.groovy [hunk 4]
```diff
@@ -54,19 +53,8 @@ class LocalTaskOutputCacheCrossBuildPerformanceTest extends AbstractCrossBuildPe
...
-                settingsFile << """
-                    buildCache {
-                        local {
-                            directory = '${cacheDir.absoluteFile.toURI()}'
-                        }
-                    }
-                """.stripIndent()
-            }
+        runner.addBuildMutator { invocationSettings ->
+            new ClearBuildCacheMutator(invocationSettings.gradleUserHome, AbstractCleanupMutator.CleanupSchedule.BUILD)
```
Comment (lptr) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/experiment/buildcache/LocalTaskOutputCacheCrossBuildPerformanceTest.groovy]: Code reuse! \o/

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractCrossBuildPerformanceTestRunner.groovy (hunk 1 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractCrossBuildPerformanceTestRunner.groovy]: I switched to `@CompileStatic`. Kotlin can infer the type after the `instanceof` check, I think the Groovy compiler can't - at least compilation failed for me.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractGradleBuildPerformanceTestRunner.groovy (hunk 2 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractGradleBuildPerformanceTestRunner.groovy]: I think an empty implementation is fine. You don't need to override `defaultSpec` in your subclass if you don't need to prepare anything in the spec.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractCrossBuildPerformanceTestRunner.groovy (hunk 1 above)
Comment (lptr) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractCrossBuildPerformanceTestRunner.groovy]: I saw this feature with Groovy first, I think it should work with `@CompileStatic` as well. This works for me in the Groovy Console in IntelliJ:

```groovy
@groovy.transform.CompileStatic
class MyType {
    static void main() {
        Object lajos = "alma"
        if (lajos instanceof String) {
            println lajos.length()
        }
    }
}
```

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractGradleBuildPerformanceTestRunner.groovy (hunk 2 above)
Comment (lptr) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractGradleBuildPerformanceTestRunner.groovy]: You are now calling `super.defaultSpec()`, which is easy to forget. I think that's more error-prone than having to implement the method.

Comment (lptr) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractGradleBuildPerformanceTestRunner.groovy]: Anyway, I don't feel very strongly about this.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractCrossBuildPerformanceTestRunner.groovy (hunk 1 above)
Comment (lptr) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractCrossBuildPerformanceTestRunner.groovy]: That said, I don't think this is important enough to require a fix. I'm still curious why it won't compile without the cast.

Comment (wolfs) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractCrossBuildPerformanceTestRunner.groovy]: It seems to work fine in IDEA, though the Groovy compiler can't compile it: https://e.grdev.net/s/adjqb3njdwwoq/console-log?task=:internalPerformanceTesting:compileGroovy#L8

**Pull request merged**